archive/imo1988_q6.lean

```diff
@@ -0,0 +1,271 @@
+  split; dsimp only,
+  { -- The first condition is not so hard. After all, c is the other root of the quadratic equation.
+    rw [H_symm, H_quad],
+    simpa using h_root, },
+  { -- For the second condition, we note that it suffices to check that c ≠ m_x.
+    suffices hc : c ≠ mx,
+    { refine lt_of_le_of_ne _ hc,
+      exact_mod_cast c_lt, },
+    -- However, recall that B(m_x) ≠ m_x + m_y.
+    -- If c = m_x, we can prove B(m_x) = m_x + m_y.
+    contrapose! hm_B₂, subst c,
+    simp [hV₁], }
+    -- Hence p' = (c, m_x) lies on the upper branch, and we are done.
+end
+
+/--Question 6 of IMO1988. If a and b are two natural numbers
+such that a*b+1 divides a^2 + b^2, show that their quotient is a perfect square.-/
+lemma imo1988_q6 {a b : ℕ} (h : (a*b+1) ∣ a^2 + b^2) :
+  ∃ d, d^2 = (a^2 + b^2)/(a*b + 1) :=
+begin
+  rcases h with ⟨k, hk⟩,
+  rw [hk, nat.mul_div_cancel_left _ (nat.succ_pos (a*b))],
+  simp only [nat.pow_two] at hk,
+  apply constant_descent_vieta_jumping a b hk (λ x, k * x) (λ x, x*x - k) (λ x y, false);
+  clear hk a b,
+  { -- We will now show that the fibres of the solution set are described by a quadratic equation.
+    intros x y, dsimp only,
+    rw [← int.coe_nat_inj', ← sub_eq_zero],
+    apply eq_iff_eq_cancel_right.2,
+    simp, ring, },
+  { -- Show that the solution set is symmetric in a and b.
+    intros x y, simp [add_comm (x*x), mul_comm x], },
+  { -- Show that the claim is true if b = 0.
+    suffices : ∀ a, a * a = k → ∃ d, d * d = k, by simpa,
+    rintros x rfl, use x },
+  { -- Show that the claim is true if a = b.
+    intros x hx,
+    suffices : k ≤ 1,
+    { rw [nat.le_add_one_iff, nat.le_zero_iff] at this,
+      rcases this with rfl|rfl,
+      { use 0, simp },
+      { use 1, simp } },
+    contrapose! hx with k_lt_one,
+    apply ne_of_lt,
+    calc x*x + x*x = x*x * 2       : by rw mul_two
+               ... ≤ x*x * k       : nat.mul_le_mul_left (x*x) k_lt_one
+               ... < (x*x + 1) * k : by apply mul_lt_mul; linarith },
+  { -- Show the descent step.
+    intros x y x_lt_y hx hxky h z h_root hV₁ hV₀,
+    split,
+    { dsimp [-sub_eq_add_neg] at *,
+      have hpos : z*z + x*x > 0,
+      { apply add_pos_of_nonneg_of_pos,
+        { apply mul_self_nonneg },
+        { apply mul_pos; exact_mod_cast hx }, },
+      have hzx : z*z + x*x = (z * x + 1) * k,
+      { rw [← sub_eq_zero, ← h_root],
+        simp, ring, },
+      rw hzx at hpos,
+      replace hpos : z * x + 1 > 0 := pos_of_mul_pos_right hpos (int.coe_zero_le k),
+      replace hpos : z * x ≥ 0 := int.le_of_lt_add_one hpos,
+      apply nonneg_of_mul_nonneg_right hpos (by exact_mod_cast hx), },
+    { contrapose! hV₀ with x_lt_z,
+      apply ne_of_gt,
+      calc z * y > x*x     : by apply mul_lt_mul'; linarith
+             ... ≥ x*x - k : sub_le_self _ (int.coe_zero_le k) }, },
+  { -- There is no base case in this application of Vieta jumping.
+    simp },
+end
+
+example {a b : ℕ} (h : a*b ∣ a^2 + b^2 + 1) :
```

Comment:
Maybe a comment explaining why you've included this example?

Reply:
Done